include/jemalloc/internal/eset.h

```diff
@@ -14,16 +14,28 @@
  * there are mutating operations.  One exception is the stats counters, which
  * may be read without any locking.
  */
+
+typedef struct eset_bin_s eset_bin_t;
+struct eset_bin_s {
+	edata_heap_t heap;
+};
+
+typedef struct eset_bin_stats_s eset_bin_stats_t;
+struct eset_bin_stats_s {
+	atomic_zu_t nextents;
+	atomic_zu_t nbytes;
+};
+
 typedef struct eset_s eset_t;
 struct eset_s {
-	/* Quantized per size class heaps of extents. */
-	edata_heap_t heaps[SC_NPSIZES + 1];
-	atomic_zu_t nextents[SC_NPSIZES + 1];
-	atomic_zu_t nbytes[SC_NPSIZES + 1];
-
 	/* Bitmap for which set bits correspond to non-empty heaps. */
 	fb_group_t bitmap[FB_NGROUPS(SC_NPSIZES + 1)];
 
+	/* Quantized per size class heaps of extents. */
+	eset_bin_t bins[SC_NPSIZES + 1];
+
+	eset_bin_stats_t bin_stats[SC_NPSIZES + 1];
```

Comment:
I thought about it, but here was my argument for the current layout: There's 3 reasonable choices:
- All summaries, then all heaps, then all stats.
- The current layout
- {summary, heap, stats}, {summary, heap, stats}, ... 

We end up touching at least the summaries for 12 different bins (with the default lg_max_active_fit). So with this layout, that means 5 cachelines (on x86_64) for bins + summaries, and 1 for stats, so 6 total. If we move in the stats, we touch 8. You could take this argument the other direction and say "well, why not split out the heap and the summary then; that touched only 3 for summaries and 1 each for stats and heaps". But that adds a potential extra cache-miss of delay until we get the likely heap address in case we miss-predict. So this layout was sort of a balance.

Really, I tried all three ways on some local benchmarks (running compilers with different versions preloaded), and couldn't find any differences, so I think any real-world differences are probably in the noise.

Reply:
Thanks for the details. I'm convinced and like the current approach.